File: Solutions/Cpp/insert_interval.cpp
/*
    [MEDIUM]
    57. Insert Interval

    Concepts:
    - array

    Stats:
        Runtime | 3 ms      [Beats 30.94%]
        Memory  | 21.52 MB  [Beats 94.25%]

*/

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.empty()) return {};

        // sort by interval start time
        std::sort(intervals.begin(), intervals.end());
        std::vector<std::vector<int>> res = {};
        res.push_back(intervals[0]);

        for (int i = 1; i < intervals.size(); i++) {
            std::vector<int>& last = res.back();

            // if current interval start before last interval end...
            if (intervals[i][0] <= last[1]) {
                // ...merge last and current
                last[1] = std::max(last[1], intervals[i][1]);
            } else {
                // create new interval
                res.push_back(intervals[i]);
            }
        }

        return res;
    }
};

File: Solutions/Rust/insert_interval.rs
/*
    [MEDIUM]
    57. Insert Interval

    Concepts:
    - array

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 2.77 MB   [Beats 36.23%]
*/
use std::cmp::max;

impl Solution {
    pub fn insert(mut intervals: Vec<Vec<i32>>, new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        intervals.push(new_interval);
        intervals.sort();

        let mut res: Vec<Vec<i32>> = Vec::new();

        for intv in intervals {
            if let Some(res_last) = res.last_mut() {
                // compare last interval's end with current interval's *start*
                if res_last[1] >= intv[0] {
                    // compare last interval's end with current interval's *end*
                    res_last[1] = max(res_last[1], intv[1]);
                    continue;
                }
            }
            // if no overlap between last and current, create new interval
            res.push(intv.clone());
        }

        return res;
    }
}